File: src/Orchestrator/services/FileArchiveService/Messages/Requests/ArchiveThingAttachments/NewThingIm.cs
using Attributes;
using Common.Models.IM;

namespace Messages.Requests;

internal class NewThingIm
{
    public required Guid SubjectId { get; init; }
    public required string Title { get; init; }
    public required string Details { get; init; }
    [ImagePath(BackingField = nameof(ImageIpfsCid))]
    public required string? ImagePath { get; init; }
    [ImagePath(BackingField = nameof(CroppedImageIpfsCid))]
    public required string? CroppedImagePath { get; init; }
    public required IEnumerable<EvidenceIm> Evidence { get; init; }
    public required IEnumerable<TagIm> Tags { get; init; }

    [BackingField]
    public required string? ImageIpfsCid { get; set; }
    [BackingField]
    public required string? CroppedImageIpfsCid { get; set; }
}

File: src/frontend/lib/thing/models/im/new_thing_im.dart
import 'dart:typed_data';

import 'package:dio/dio.dart';

import '../../../general/models/im/tag_im.dart';
import 'evidence_im.dart';

class NewThingIm {
  final String subjectId;
  final String title;
  final String details;
  final String? imageExt;
  final Uint8List? imageBytes;
  final String? croppedImageExt;
  final Uint8List? croppedImageBytes;
  final List<EvidenceIm> evidence;
  final List<TagIm> tags;

  NewThingIm({
    required this.subjectId,
    required this.title,
    required this.details,
    required this.imageExt,
    required this.imageBytes,
    required this.croppedImageExt,
    required this.croppedImageBytes,
    required this.evidence,
    required this.tags,
  });

  FormData toFormData() {
    // @@??: Is order of sections guaranteed?
    var map = <String, dynamic>{
      'subjectId': subjectId,
      'title': title,
      'details': details,
      'evidence': evidence.map((e) => e.url).join('|'),
      'tags': tags.map((t) => t.id).join('|'),
    };

    if (imageBytes != null && croppedImageBytes != null) {
      map['image'] = MultipartFile.fromBytes(
        imageBytes!,
        filename: 'image.$imageExt',
      );
      map['croppedImage'] = MultipartFile.fromBytes(
        croppedImageBytes!,
        filename: 'cropped_image.$croppedImageExt',
      );
    }

    return FormData.fromMap(map);
  }
}
